File: _downloads/314bc9f7a912ca85fdea024afb8a67cf/camera.cpp
#include <iostream>
#include <glm/glm.hpp>
#include <glm/gtx/io.hpp>

#include "camera.hpp"

Camera::Camera(const glm::vec3 position)
{
    this->position = position;
    front = glm::vec3(0.0f, 0.0f, -1.0f);
    worldUp = glm::vec3(0.0f, 1.0f, 0.0f);
}

glm::mat4 Camera::getViewMatrix()
{
    return view;
}

glm::mat4 Camera::getProjectionMatrix()
{
    return projection;
}

void Camera::calculateMatrices(GLFWwindow *window, const float deltaTime)
{
    // Keyboard inputs
    if (glfwGetKey(window, GLFW_KEY_W))
        position += front * deltaTime * speed;
    
    if (glfwGetKey(window, GLFW_KEY_S))
        position -= front * deltaTime * speed;
    
    if (glfwGetKey(window, GLFW_KEY_A))
        position -= right * deltaTime * speed;
    
    if (glfwGetKey(window, GLFW_KEY_D))
        position += right * deltaTime * speed;
    
    // Get mouse cursor position
    double xPos, yPos;
    glfwGetCursorPos(window, &xPos, &yPos);
    glfwSetCursorPos(window, 1024/2, 768/2);
    
    // Update yaw and pitch angles
    yaw += mouseSpeed * deltaTime * float(xPos - 1024/2);
    pitch += mouseSpeed * deltaTime * float(yPos - 768/2);

    // Calculate front vector
    front = glm::normalize(glm::vec3(cos(pitch) * sin(yaw) , sin(pitch), -cos(yaw) * cos(pitch)));
    
    // Calculate view matrix
    view = glm::lookAt(position, position + front, worldUp);
    
    // Update camera vectors
    right.x = view[0][0],  right.y = view[1][0],  right.z = view[2][0];
    up.x    = view[0][1],  up.y    = view[1][1],  up.z    = view[2][1];
    
    // Calculate projection matrix
    projection = glm::perspective(fov, aspect, near, far);
}

File: _webgl/Lab 6 - 3D Worlds/camera.js
class Camera {

  constructor() {

    // Camera vectors
    this.eye     = [0, 0, 0];
    this.worldUp = [0, 1, 0];
    this.front   = [0, 0, -1];
    this.right   = [1, 0, 0];
    this.up      = [0, 1, 0];

    // Projection settings
    this.fov    = 45 * Math.PI / 180;
    this.aspect = 800 / 600;
    this.near   = 0.1;
    this.far    = 1000;
  }

  update() {
    this.right = normalize(cross(this.front, this.worldUp));
    this.up    = normalize(cross(this.right, this.front));
  }

  getViewMatrix() {
    return new Mat4().set([
        this.right[0], this.up[0], -this.front[0], 0,
        this.right[1], this.up[1], -this.front[1], 0,
        this.right[2], this.up[2], -this.front[2], 0,
        -dot(this.eye, this.right),
        -dot(this.eye, this.up),
            dot(this.eye, this.front),
        1
    ]);
  }

  getOrthographicMatrix(left, right, bottom, top, near, far) {
    const rl = 1 / (right - left);
    const tb = 1 / (top - bottom);
    const fn = 1 / (far - near);

    return new Mat4().set([
        2 * rl, 0, 0, 0,
        0, 2 * tb, 0, 0,
        0, 0, -2 * fn, 0,
        -(right + left) * rl,
        -(top + bottom) * tb,
        -(far + near) * fn,
        1
    ]);
  }

  getPerspectiveMatrix() {
    const t  = this.near * Math.tan(this.fov / 2);
    const r  = this.aspect * t;
    const fn = 1 / (this.far - this.near);

    return new Mat4().set([
      this.near / r, 0, 0, 0,
      0, this.near / t, 0, 0,
      0, 0, -(this.far + this.near) * fn, -1,
      0, 0, -2 * this.far * this.near * fn, 0
    ]);
  }
}